index.bs

```diff
@@ -4648,71 +4659,156 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
    </dd>
 </dl>
 
-<div algorithm="remote end steps for browsingContext.setViewport">
+<div algorithm>
+To <dfn>set device pixel ratio override</dfn> given |navigable| and |device pixel ratio|:
 
-The [=remote end steps=] with |command parameters| are:
+1. If |device pixel ratio| is not null:
 
-1. Let |navigable id| be the value of the <code>context</code> field of |command
-   parameters|.
+   1. For [=/document=] currently loaded in a specified |navigable|:
 
-1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
-   |navigable id|.
+      1. When the [=select an image source from a source set=] steps are run, act as if
+         the implementation's pixel density was set to |device pixel ratio| when selecting an image.
 
-1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
-   [=error code=] [=invalid argument=].
+      1. For the purposes of the [=resolution media feature=], act as if
+         the implementation's resolution is |device pixel ratio| dppx scaled by the page zoom.
+
+   1. [=map/Set=] [=device pixel ratio overrides=][|navigable|] to |device pixel ratio|.
+
+      Note: This will take an effect because of the patch of [[#patchs-determine-the-device-pixel-ratio]].
+
+1. Otherwise:
+
+   1. For [=/document=] currently loaded in a specified |navigable|:
+
+      1. When the [=select an image source from a source set=] steps are run, use the implementation's default behavior,
+         without any changes made by previous invocations of these steps.
+
+      1. For the purposes of the [=resolution media feature=], use the implementation's default behavior,
+         without any changes made by previous invocations of these steps.
+
+   1. [=map/Remove=] |navigable| from [=device pixel ratio overrides=].
+
+1. Run [=evaluate media queries and report changes=] for [=/document=] currently loaded
+   in a specified |navigable|.
+
+</div>
+
+
+<div algorithm>
+To <dfn>set viewport</dfn> given |navigable| and |viewport|:
+
+1. If |viewport| is not null, set the width of |navigable|'s [=layout
+   viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
+   set the height of the |navigable|'s [=layout viewport=] to be the
+   <code>height</code> of |viewport| in CSS pixels.
+
+1. Otherwise, set the |navigable|'s [=layout viewport=] to the
+   implementation-defined default.
+
+</div>
+
+<div algorithm="set viewport when a navigable is created">
+
+After creating a document in a new [=/navigable=] |navigable| and
+before the [=run WebDriver BiDi preload scripts=] algorithm is invoked:
+
+TODO: Move it as a hook in the html spec instead.
```

Comment:
We already have a "WebDriver BiDi navigable created" hook. I believe it's the right moment to set the viewports, and we don't need an additional hook.

Reply:
But at that moment we don't have (guaranteed) a document created, which we need to apply viewport settings to

Reply:
indeed